daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PState.scala

```diff
@@ -304,23 +302,38 @@ final class PState private (
   def removeVariableInstance(vrd: VariableRuntimeData): Unit = {
     variableMap.removeVariableInstance(vrd)
   }
+  
+  def withPointOfUncertainty[B](pouID: String, context: RuntimeData)(func: PState.Mark => B): B =
+    macro PointOfUncertaintyMacros.withPointOfUncertainty[PState.Mark, B]
 
-  def pushPointOfUncertainty: Unit = {
-    // threadCheck()
-    discriminatorStack.push(false)
-    changedVariablesStack.push(mutable.MutableList[GlobalQName]())
+  // must be public, but should only be used by the withPointOfUncertainty macro
+  def createPointOfUncertainty(pouID: String, context: RuntimeData): PState.Mark = {
+    val pou = mark(pouID, context)
+    pointsOfUncertainty.push(pou)
+    pou
   }
 
-  def popPointOfUncertainty: Unit = {
-    // threadCheck()
-    discriminatorStack.pop
-    changedVariablesStack.pop
+  def resetToPointOfUncertainty(pou: PState.Mark): Unit = {
+    val pouPop = pointsOfUncertainty.pop
+    Assert.invariant(pou == pouPop)
+    reset(pouPop)
   }
 
-  def setDiscriminator(disc: Boolean): Unit = {
-    // threadCheck()
-    discriminatorStack.pop()
-    discriminatorStack.push(disc)
+  def discardPointOfUncertainty(pou: PState.Mark): Unit = {
+    val pouPop = pointsOfUncertainty.pop
+    Assert.invariant(pou == pouPop)
+    discard(pou)
+  }
+
+  def resolvePointOfUncertainty(): Unit = {
+    if (!pointsOfUncertainty.isEmpty) {
+      val pouPop = pointsOfUncertainty.pop
+      discard(pouPop)
+    }
+  }
+
+  def isPointOfUncertaintyResolved(pou: PState.Mark): Boolean = {
```

Comment:
Seems odd to me. I expected this to search down the stack, it's only checking the top. Can you add scaladoc explaining why that's sufficient? The principles of operation here are quite different now, because the parsers manipulate PoUs via a local variable now, not manipulating or searching a stack at all.  So the stack is now just to identify the current unresolved PoU. This all needs to be in code comments/scaladoc not just in the intro to this PR.

Reply:
The quick description is: parsers push PoU's onto the top of the stack when one exists. So the the current in-scope PoU is always at the top of the stack. Discriminators/initated content/etc. resolve whatever is on the top of the stack. This means resolve a PoU changes the in-scope PoU, allowing multiple discriminators to resolve multiple PoUs down the stack. When parsers that created PoU's wind down, two cases are possible:
1. The parers PoU is at the top of the stack, which means nothing resolved it
2. Some previous PoU is at the top of the stack, which can only happen when a discriminator/initated content/etc. popped it off, meaning it was resolved. So this isPoUResolved function really just needs to see if the PoU the parser created is at the top of the stack.

I'll add some doc to all these things to make this clear.